DLSv2/Core/Wrappers.cs

```diff
@@ -0,0 +1,162 @@
+                    else
+                    {
+                        Enabled = !fromToggle || previousStatus;
+                        Index = newIndex;
+                    }
+                    break;
+                }
+                case ControlGroupInstance<AudioControlGroup> audioCG:
+                    if (previousStatus == false && !fromToggle)
+                    {
+                        if (Index == 0 || Index == audioCG.BaseControlGroup.Modes.Count - 1)
+                        {
+                            Enabled = true;
+                            Index = 0;
+                            return;
+                        }
+                    }
+                    if (newIndex >= audioCG.BaseControlGroup.Modes.Count)
+                    {
+                        Enabled = fromToggle ? previousStatus : !cycleOnly;
+                        Index = 0;
+                    }
+                    else
+                    {
+                        Enabled = !fromToggle || previousStatus;
+                        Index = newIndex;
+                    }
+                    break;
+            }
+        }
+
+        public void MoveToPrevious(bool cycleOnly = false)
+        {
+            var prevIndex = Index;
+            var newIndex = prevIndex - 1;
+
+            switch (this)
+            {
+                case ControlGroupInstance<ControlGroup> lightCG:
+                {
+                    if (Enabled == false)
+                    {
+                        if (Index == 0)
+                        {
+                            Enabled = true;
+                            Index = lightCG.BaseControlGroup.Modes.Count - 1;
+                            return;
+                        }
+                    }
+                    if (newIndex < 0)
+                    {
+                        Enabled = false;
+                        Index = 0;
+                    }
+                    else
+                    {
+                        Enabled = true;
+                        Index = newIndex;
+                    }
+                    break;
+                }
+                case ControlGroupInstance<AudioControlGroup> audioCG:
+                    if (Enabled == false)
+                    {
+                        if (Index == 0)
+                        {
+                            Enabled = true;
+                            Index = audioCG.BaseControlGroup.Modes.Count - 1;
+                            return;
+                        }
+                    }
+                    if (newIndex < 0)
+                    {
+                        Enabled = !cycleOnly;
+                        Index = cycleOnly ? 0 : audioCG.BaseControlGroup.Modes.Count - 1;
+                    }
+                    else
+                    {
+                        Enabled = true;
+                        Index = newIndex;
+                    }
+                    break;
+            }
```

Comment:
This logic seems to be the same for audio control groups and light control groups. Likewise in MoveToNext. It might make sense to have a base ControlGroup class that both light and audio control groups inherit from. Then ControlGroupInstance can have `<T>` be forced to use a `BaseControlGroup` (`ControlGroupInstance<T> where T : BaseControlGroup` so you can use properties of `T` in methods like next/prev.

Reply:
Audio and Light control groups differ on what happens on if newIndex < 0.

Reply:
You could still have the same base class so the types are cleaner, and then implement a different logic for just that line